### ISSUE TITLE
Numerous issues with 0.8.0-rc1 redhat RPM spec file

### ISSUE THREAD

**Issue by stewartadam** [zfsonlinux/zfs#7882]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Fedora
Distribution Version	| 27
Linux Kernel	| 4.17.17-100.fc27.x86_64
Architecture	| x86_64
ZFS Version	| 0.8.0rc1

### Describe the problem you're observing
Trying to build 0.8.0rc1 for Fedora (RPM) and encountering several issues. Overall scope to the issue is that as a user, I would expect `rpmbuild -ba zfs.spec` or `rpmbuild --rebuild zfs-foo.src.rpm` to work out of the box without having to invoke any magic arguments:
1. Several Makefiles were missing libzfs_core.la from LDADD resulting in undefined symbols at link time in `tests/` and `cmd/`:
    ```
    Making all in libzfs
    make[6]: Entering directory '/home/stewartadam/devel/rpmbuild/BUILD/zfs-0.8.0/tests/zfs-tests/tests/functional/libzfs'
      CC       many_fds.o
      CCLD     many_fds
    ../../../../../lib/libzfs/.libs/libzfs.so: undefined reference to `lzc_pool_checkpoint_discard'
    ../../../../../lib/libzfs/.libs/libzfs.so: undefined reference to `lzc_remap'
    ../../../../../lib/libzfs/.libs/libzfs.so: undefined reference to `lzc_reopen'
    ../../../../../lib/libzfs/.libs/libzfs.so: undefined reference to `lzc_channel_program_nosync'
    ../../../../../lib/libzfs/.libs/libzfs.so: undefined reference to `lzc_pool_checkpoint'
    collect2: error: ld returned 1 exit status
    make[6]: *** [Makefile:692: many_fds] Error 1
    make[6]: Leaving directory '/home/stewartadam/devel/rpmbuild/BUILD/zfs-0.8.0/tests/zfs-tests/tests/functional/libzfs'
    make[5]: *** [Makefile:681: all-recursive] Error 1
    ```
    This may be a race condition as I've nuked my compile directory and re-extracted the upstream tarball several times to make sure I can repro - it's failed three times, and built twice. I've used various methods to try and build, per below.

2. `%{_generatordir}` for systemd is undefined; the current macro appears to be `%{_systemdgeneratordir}`: [see here](https://github.com/systemd/systemd/blob/master/src/core/macros.systemd.in)
3. pyzfs files are built and installed regardless of pyzfs define, so if `--define '_pyzfs 1'` is not used then RPM complains of unpackaged files in the python site-packages folder
4. There doesn't appear to be a way to enable pyzfs short of manually defining like above, we should consider a conditional flag using `--with` instead

### Describe how to reproduce the problem
```
cd ~/rpmbuild/SOURCES
wget https://github.com/zfsonlinux/zfs/releases/download/zfs-0.8.0-rc1/zfs-0.8.0-rc1.tar.gz
tar xfz zfs-0.8.0-rc1.tar.gz
cd zfs-0.8.0
./configure --with-config=user
cd rpm/redhat
sed -i -e 's/%{name}-%{version}.tar.gz/%{name}-%{version}-rc1.tar.gz/' zfs.spec
cp zfs.spec ../../SPECS
rpmbuild -ba zfs.spec
```

Same result if I were to build `make srpm`, `rpmbuild --rebuild $SRC.rpm`, or `make pkg-utils rpm-dkms`.

**Comment by stewartadam** [comment]
I've started committing some fixes [here](https://github.com/zfsonlinux/zfs/compare/master...stewartadam:rpmfixes/0.8.0rc1)

* Links against `libzfs_core.la` when applicable to avoid missing symbols
* Uses upstream `%{_systemdgeneratordir}` macro instead of derived `%{_generatordir}` macro
* Switches to a `--with` flag for pyzfs and initramfs sub-packages instead of manual defines

Item 2 may be expected given the m4 macros I'm seeing in the repo, but I don't know what that design choice was made instead of using the native macro - it breaks building the RPMs outside of `make pkg-utils`. Please advise on which direction you'd like to take.

**Comment by stewartadam** [comment]
As well, is it intended for RPM builds to use `./configure --with-config=user` (which skips the majority of feature verifications seen when executing only `./configure` on the same system)?

**Comment by loli10K** [comment]
Duplicate of https://github.com/zfsonlinux/zfs/issues/6539

**Comment by loli10K** [comment]
Duplicate of https://github.com/zfsonlinux/zfs/issues/7567

**Comment by stewartadam** [comment]
Thanks, I'll continue discussion for the issues 1 and 2 on the specific threads.

What about the `--with-config` issue? Is `user` expected to be used when compiling the userspace tools, `kernel` for the kmod?

**Comment by loli10K** [comment]
I only linked those issues so we don't forget to close them when they get fixed (keeping track of ~1K issues is not an easy task).

> What about the --with-config issue? Is user expected to be used when compiling the userspace tools, kernel for the kmod?

Yes, that is my understanding.

**Comment by behlendorf** [comment]
@stewartadam thanks for looking in to these packaging issue.  It would be great if you could open a PR with the propsed fixes so we can get then reviewed and you some feedback.

> What about the --with-config issue? Is user expected to be used when compiling the userspace tools, kernel for the kmod?

Yes, exactly.  The option allows us to build either the user space or kernel space code independently to be packaged.  And as you noticed ` --with-config=user` disables all of the kernel specific configure checks.  These checks are only required to build the kmod and the user space tools and libraries must be decoupled from this.  This was a single version of the user space packages can be built and installed along with kmod packages for all the kernels which may be installed.  This could definitely be better documented as part of configure.

**Comment by tonyhutter** [comment]
@stewartadam I took your generators macro fix and opened a PR for it here: https://github.com/zfsonlinux/zfs/pull/8107.  It will fix one aspect of this issue:

> 2. %{_generatordir} for systemd is undefined; the current macro appears to be %{_systemdgeneratordir}: see here

**Comment by behlendorf** [comment]
The issues listed here should be resolved in 0.8.0-rc4.
